exec: panic when selecting from crdb_internal.zones

When `experimental_vectorize` is enabled at the cluster level, selecting from `crdb_internal.zones` panics.
```
root@127.0.0.1:65524/defaultdb> SET CLUSTER SETTING sql.defaults.experimental_vectorize = on;
SET CLUSTER SETTING

Time: 2.896ms

root@127.0.0.1:65524/defaultdb> SELECT * FROM crdb_internal.zones;
E190702 16:28:07.539314 740 sql/conn_executor.go:713  [n1,client=127.0.0.1:65526,user=root] a SQL panic has occurred while executing "SELECT * FROM crdb_internal.zones": expected *DInt, found tree.dNull
E190702 16:28:07.539387 740 util/log/crash_reporting.go:207  [n1,client=127.0.0.1:65526,user=root] a panic has occurred!
panic: expected *DInt, found tree.dNull [recovered]
	panic: panic while executing 1 statements: SELECT * FROM _._; caused by expected *DInt, found tree.dNull

goroutine 740 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc000cca380, 0x75f07e0, 0xc000a11580, 0x6a95f60, 0xc0003b7340)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:727 +0x330
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc000cca380, 0x75f07e0, 0xc000a11580)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:440 +0x61
panic(0x6a95f60, 0xc0003b7340)
	/usr/local/Cellar/go/1.12.6/libexec/src/runtime/panic.go:522 +0x1b5
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.MustBeDInt(0x75f7d60, 0x9707c20, 0x9707c20)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/datum.go:617 +0xd9
github.com/cockroachdb/cockroach/pkg/sql.(*planner).getAllNames(0xc000cca6d0, 0x75f08a0, 0xc0014fa600, 0x10, 0x400e999, 0xc001692f90)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/crdb_internal.go:1835 +0x16f
github.com/cockroachdb/cockroach/pkg/sql.glob..func53(0x75f08a0, 0xc0014fa600, 0xc000cca6d0, 0xc00041ab20, 0xc00175dc20, 0xc00175dbd0, 0xc000630001)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/crdb_internal.go:1866 +0x5d
github.com/cockroachdb/cockroach/pkg/sql.virtualDefEntry.getPlanInfo.func1(0x75f08a0, 0xc0014fa600, 0xc000cca6d0, 0x6dd8709, 0x9, 0xc0002da5c0, 0xc00063a0a0, 0xc0007290e0, 0x100)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_schema.go:272 +0x38e
github.com/cockroachdb/cockroach/pkg/sql.(*execFactory).ConstructVirtualScan.func1(0x75f08a0, 0xc0014fa600, 0xc000cca6d0, 0x1, 0xc0014fa5d0, 0x0, 0x0)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt_exec_factory.go:135 +0x5e
github.com/cockroachdb/cockroach/pkg/sql.(*delayedNode).startExec(0xc001515840, 0x75f08a0, 0xc0014fa600, 0xc001862800, 0xc000cca6d0, 0x1, 0xc0011a5b30)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/delayed.go:58 +0x57
github.com/cockroachdb/cockroach/pkg/sql.startExec.func2(0x6de271b, 0xd, 0x75f1d60, 0xc001515840, 0xc0000b1bc0, 0xc0011a5b40)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:497 +0x55
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal.func1(0xc00179b7a0, 0x6de271b, 0xd, 0x75f1d60, 0xc001515840)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:142 +0x5d
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal(0xc00179b7a0, 0x75f1d60, 0xc001515840, 0x6de271b, 0xd)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:624 +0x2a8
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visit(0xc00179b7a0, 0x75f1d60, 0xc001515840, 0x0, 0x60)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:109 +0x8d
github.com/cockroachdb/cockroach/pkg/sql.walkPlan(...)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:73
github.com/cockroachdb/cockroach/pkg/sql.startExec(0x75f08a0, 0xc0014fa600, 0xc001862800, 0xc000cca6d0, 0x75f1d60, 0xc001515840, 0x75abfe0, 0x970a660)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:500 +0x286
github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Start(0xc000634500, 0x75f08a0, 0xc0014fa600, 0x970a660, 0x68ac3a0)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:117 +0xd6
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ProcessorBase).Run(0xc000634500, 0x75f08a0, 0xc0014fa600)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go:792 +0x52
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).Run(0xc00000cd20, 0x75f08a0, 0xc0014fa600, 0x6f49358, 0x0, 0x0)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:654 +0x1f2
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc0001eafc0, 0xc00179b6e0, 0xc001675710, 0xc0011a68b0, 0xc00092e2c0, 0xc000cca7a8, 0x0)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:296 +0x35c
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0xc0001eafc0, 0x75f08a0, 0xc0014fa420, 0xc000cca7a8, 0xc00179b6e0, 0xc001675710, 0x75f1d60, 0xc001515840, 0xc00092e2c0)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:860 +0x208
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc000cca380, 0x75f08a0, 0xc0014fa420, 0xc000cca6d0, 0x3, 0xbb92940, 0xc000443d90, 0xc0011a6b00, 0x0, 0x0, ...)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:902 +0x370
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc000cca380, 0x75f08a0, 0xc0014fa420, 0xc000cca6d0, 0xbb92940, 0xc000443d90, 0x0, 0x0)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:734 +0x6e5
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc000cca380, 0x75f08a0, 0xc0014fa420, 0x75f5ca0, 0xc0018610c0, 0xc0008d6170, 0x21, 0x0, 0x1, 0x0, ...)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:417 +0xb64
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc000cca380, 0x75f08a0, 0xc0014fa420, 0x75f5ca0, 0xc0018610c0, 0xc0008d6170, 0x21, 0x0, 0x1, 0x0, ...)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:99 +0x4ec
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc000cca380, 0x75f07e0, 0xc001515740, 0x0, 0x0)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1204 +0x1ab0
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc000cca380, 0x75f07e0, 0xc000a11580, 0xc000507540, 0x5400, 0x15000, 0xc0005075d8, 0xc0002c98a0, 0x0, 0x0)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1140 +0x1a3
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc0001b9b00, 0x75f07e0, 0xc000a11580, 0xc000cca380, 0x5400, 0x15000, 0xc0005075d8, 0xc0002c98a0, 0x0, 0x0)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:442 +0xce
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1(0xc000347d39, 0xc0003091e0, 0x75f07e0, 0xc000a11580, 0xc0002c98a0, 0xc0001b9b00, 0xc000b9e000, 0x75f4220, 0xc0003091c0, 0xc001020300, ...)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:580 +0x21c
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:509 +0x17b
```

--- a/docs/tech-notes/encoding.md
+++ b/docs/tech-notes/encoding.md
@@ -126,6 +126,13 @@ the KV value has other data, to guarantee that primary keys appear in at
 least one KV pair. (Even if there are other column families, their KV
 pairs may be suppressed; see subsection NULL below.)
 
+Note that in system tables that use multiple column families, such as
+system.zones or system.namespace, there may not be any sentinel KV pair at all.
+This is because of the fact that the database writes to these system tables
+using raw KV puts and does not include the logic to write a sentinel KV. KV
+decoding code that needs to understand system tables must be aware of this
+possibility.
+
 ### Single-column column families
 
 Before column families (i.e., in format version 1), non-sentinel KV keys

--- a/pkg/sql/colencoding/key_encoding.go
+++ b/pkg/sql/colencoding/key_encoding.go
@@ -16,6 +16,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/exec/coldata"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlbase"
 	"github.com/cockroachdb/cockroach/pkg/sql/types"
+	"github.com/cockroachdb/cockroach/pkg/util"
 	"github.com/cockroachdb/cockroach/pkg/util/encoding"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/errors"
@@ -63,7 +64,8 @@ func DecodeIndexKeyToCols(
 			}
 
 			length := int(ancestor.SharedPrefixLen)
-			key, err = DecodeKeyValsToCols(vecs, idx, indexColIdx[:length], types[:length], colDirs[:length], key)
+			key, err = DecodeKeyValsToCols(vecs, idx, indexColIdx[:length], types[:length], colDirs[:length],
+				nil /* unseen */, key)
 			if err != nil {
 				return nil, false, err
 			}
@@ -94,7 +96,7 @@ func DecodeIndexKeyToCols(
 		}
 	}
 
-	key, err = DecodeKeyValsToCols(vecs, idx, indexColIdx, types, colDirs, key)
+	key, err = DecodeKeyValsToCols(vecs, idx, indexColIdx, types, colDirs, nil /* unseen */, key)
 	if err != nil {
 		return nil, false, err
 	}
@@ -115,6 +117,9 @@ func DecodeIndexKeyToCols(
 // result to the idx'th slot of the input slice of exec.ColVecs. If the
 // directions slice is nil, the direction used will default to
 // encoding.Ascending.
+// If the unseen int set is non-nil, upon decoding the column with ordinal i,
+// i will be removed from the set to facilitate tracking whether or not columns
+// have been observed during decoding.
 // See the analog in sqlbase/index_encoding.go.
 func DecodeKeyValsToCols(
 	vecs []coldata.Vec,
@@ -122,6 +127,7 @@ func DecodeKeyValsToCols(
 	indexColIdx []int,
 	types []types.T,
 	directions []sqlbase.IndexDescriptor_Direction,
+	unseen *util.FastIntSet,
 	key []byte,
 ) ([]byte, error) {
 	for j := range types {
@@ -135,6 +141,9 @@ func DecodeKeyValsToCols(
 			// Don't need the coldata - skip it.
 			key, err = skipTableKey(&types[j], key, enc)
 		} else {
+			if unseen != nil {
+				unseen.Remove(i)
+			}
 			key, err = decodeTableKeyToCol(vecs[i], idx, &types[j], key, enc)
 		}
 		if err != nil {
@@ -146,7 +155,7 @@ func DecodeKeyValsToCols(
 
 // decodeTableKeyToCol decodes a value encoded by EncodeTableKey, writing the result
 // to the idx'th slot of the input exec.Vec.
-// See the analog, DecodeTableKey, in
+// See the analog, DecodeTableKey, in sqlbase/column_type_encoding.go.
 func decodeTableKeyToCol(
 	vec coldata.Vec, idx uint16, valType *types.T, key []byte, dir sqlbase.IndexDescriptor_Direction,
 ) ([]byte, error) {

--- a/pkg/sql/logictest/testdata/logic_test/vectorize
+++ b/pkg/sql/logictest/testdata/logic_test/vectorize
@@ -434,3 +434,13 @@ query ITI
 SELECT * FROM system.namespace LIMIT 1
 ----
 0  defaultdb  50
+
+# Regression test for issue with fetching from unique indexes with embedded
+# nulls.
+statement ok
+CREATE TABLE t38753 (x INT PRIMARY KEY, y INT, UNIQUE INDEX (y)); INSERT INTO t38753 VALUES (0, NULL)
+
+query II
+SELECT * FROM t38753 ORDER BY y;
+----
+0  NULL

--- a/pkg/sql/logictest/testdata/logic_test/vectorize
+++ b/pkg/sql/logictest/testdata/logic_test/vectorize
@@ -422,3 +422,15 @@ INSERT INTO t38626 VALUES (1, 'hello')
 
 statement ok
 EXPERIMENTAL SCRUB TABLE t38626
+
+# Regression test for issue with reading from system tables that have no
+# sentinel keys.
+query T
+SELECT "hashedPassword" FROM system.users LIMIT 1
+----
+·
+
+query ITI
+SELECT * FROM system.namespace LIMIT 1
+----
+0  defaultdb  50

--- a/pkg/sql/row/cfetcher.go
+++ b/pkg/sql/row/cfetcher.go
@@ -18,6 +18,7 @@ import (
 	"strings"
 
 	"github.com/cockroachdb/cockroach/pkg/internal/client"
+	"github.com/cockroachdb/cockroach/pkg/keys"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/sql/colencoding"
 	"github.com/cockroachdb/cockroach/pkg/sql/exec/coldata"
@@ -578,7 +579,7 @@ func (rf *CFetcher) NextBatch(ctx context.Context) (coldata.Batch, error) {
 		case stateDecodeFirstKVOfRow:
 			if rf.mustDecodeIndexKey {
 				if debugState {
-					log.Infof(ctx, "Decoding key %s", rf.machine.nextKV.Key)
+					log.Infof(ctx, "Decoding first key %s", rf.machine.nextKV.Key)
 				}
 				key, matches, err := colencoding.DecodeIndexKeyToCols(
 					rf.machine.colvecs,
@@ -606,10 +607,23 @@ func (rf *CFetcher) NextBatch(ctx context.Context) (coldata.Batch, error) {
 				}
 				prefix := rf.machine.nextKV.Key[:len(rf.machine.nextKV.Key)-len(key)]
 				rf.machine.lastRowPrefix = prefix
+			} else {
+				// If mustDecodeIndexKey was false, we can't possibly have an
+				// interleaved row on our hands, so we can figure out our row prefix
+				// without parsing any keys by using GetRowPrefixLength.
+				prefixLen, err := keys.GetRowPrefixLength(rf.machine.nextKV.Key)
+				if err != nil {
+					return nil, err
+				}
+				rf.machine.lastRowPrefix = rf.machine.nextKV.Key[:prefixLen]
+			}
+			familyID, err := rf.getCurrentColumnFamilyID()
+			if err != nil {
+				return nil, err
 			}
 			rf.machine.remainingValueColsByIdx.CopyFrom(rf.table.neededValueColsByIdx)
 			// Process the current KV's value component.
-			if _, _, err := rf.processValue(ctx, sqlbase.FamilyID(0)); err != nil {
+			if _, _, err := rf.processValue(ctx, familyID); err != nil {
 				return nil, err
 			}
 			if rf.table.isSecondaryIndex || len(rf.table.desc.Families) == 1 {
@@ -657,6 +671,10 @@ func (rf *CFetcher) NextBatch(ctx context.Context) (coldata.Batch, error) {
 			// prefix and indicate that it needs decoding.
 			rf.machine.nextKV = kv
 
+			if debugState {
+				log.Infof(ctx, "Decoding next key %s", rf.machine.nextKV.Key)
+			}
+
 			// TODO(jordan): optimize this prefix check by skipping span prefix.
 			if !bytes.HasPrefix(kv.Key, rf.machine.lastRowPrefix) {
 				// The kv we just found is from a different row.
@@ -677,14 +695,12 @@ func (rf *CFetcher) NextBatch(ctx context.Context) (coldata.Batch, error) {
 				continue
 			}
 
-			var id uint64
-			_, id, err = encoding.DecodeUvarintAscending(key)
-			familyID := sqlbase.FamilyID(id)
+			familyID, err := rf.getCurrentColumnFamilyID()
 			if err != nil {
-				return nil, scrub.WrapError(scrub.IndexKeyDecodingError, err)
+				return nil, err
 			}
 
-			// Process the current KV'rf value component.
+			// Process the current KV's value component.
 			if _, _, err := rf.processValue(ctx, familyID); err != nil {
 				return nil, err
 			}
@@ -1043,4 +1059,24 @@ func (rf *CFetcher) fillNulls() error {
 // The RangeInfo's are deduped and not ordered.
 func (rf *CFetcher) GetRangesInfo() []roachpb.RangeInfo {
 	return rf.fetcher.getRangesInfo()
+}
+
+// getCurrentColumnFamilyID returns the column family id of the key in
+// rf.machine.nextKV.Key.
+func (rf *CFetcher) getCurrentColumnFamilyID() (sqlbase.FamilyID, error) {
+	// If the table only has 1 column family, and its ID is 0, we know that the
+	// key has to be the 0th column family.
+	if rf.table.maxColumnFamilyID == 0 {
+		return 0, nil
+	}
+	// The column family is encoded in the final bytes of the key. The last
+	// byte of the key is the length of the column family id encoding
+	// itself. See encoding.md for more details, and see MakeFamilyKey for
+	// the routine that performs this encoding.
+	var id uint64
+	_, id, err := encoding.DecodeUvarintAscending(rf.machine.nextKV.Key[len(rf.machine.lastRowPrefix):])
+	if err != nil {
+		return 0, scrub.WrapError(scrub.IndexKeyDecodingError, err)
+	}
+	return sqlbase.FamilyID(id), nil
 }

--- a/pkg/sql/row/cfetcher.go
+++ b/pkg/sql/row/cfetcher.go
@@ -840,6 +840,7 @@ func (rf *CFetcher) processValue(
 				table.extraValColOrdinals,
 				table.extraTypes,
 				nil,
+				&rf.machine.remainingValueColsByIdx,
 				valueBytes,
 			)
 			if err != nil {

